### PR TITLE
Noman/authentication rework

### PR DESCRIPTION
**Pull request by nomanSanjari**
We've now gotten rid of any OTP functionality. Upon attempting to register, validity of the institution and whether a user with that email exists or not is checked. If the request is valid, that is, the institution is valid and the user is new, then a JWT is minted and they are sent that JWT as a link to their email. Upon clicking, the request is received by the backend and the token is verified(de-minted) back to the original payload. A final series of validity and uniqueness checks are performed, then the user is added to the database.

### CODE REVIEW

File: backend/controllers/user.controller.ts [hunk 1]
```diff
@@ -65,155 +51,101 @@ export const createNewUser = async (
...
-    const otp = otpGenerator.generate(6, {
-      upperCaseAlphabets: false,
-      lowerCaseAlphabets: false,
-      specialChars: false,
-    });
-
-    await prisma.user.update({
-      where: {
-        email: email,
+      process.env.JWT_SECRET ?? "testSecret",
```
Comment (AarshShah9) [inline backend/controllers/user.controller.ts:69]: Please use the env validator instead of going straight to the process.

File: backend/controllers/user.controller.ts [hunk 2]
```diff
@@ -236,84 +168,38 @@ export const loginUser = async (
...
+        {
+          ID: existingUser.id,
+          institutionID: existingUser.institutionId,
+          username: existingUser.username,
+          firstName: existingUser.firstName,
+          lastName: existingUser.lastName,
+          email: existingUser.email,
+          password: existingUser.password,
+        },
+        process.env.JWT_SECRET ?? "testSecret",
```
Comment (AarshShah9) [inline backend/controllers/user.controller.ts:187]: same thing here

File: backend/prisma/schema.prisma [hunk 3]
```diff
@@ -66,16 +66,11 @@ model User {
...
-
-  otp        String? @default("")
-  jwt        String? @default("")
-  isVerified Boolean @default(false) @map("is_verified")
 
   profilePic String? @unique @map("profile_pic")
 
   institution   Institution @relation(fields: [institutionId], references: [id], onDelete: Cascade)
-  institutionId String      @map("institution_id")
+  institutionId String
```
Comment (AarshShah9) [inline backend/prisma/schema.prisma:73]: why are you getting rid of the map

File: backend/prisma/schema.prisma [hunk 4]
```diff
@@ -46,7 +46,7 @@ enum AppPermissionName {
 model Institution {
   id     String @id @default(uuid())
   name   String
-  domain String
+  domain String // fix double entry issue
```
Comment (AarshShah9) [inline backend/prisma/schema.prisma:49]: just make this an @unique no?

File: backend/controllers/user.controller.ts [hunk 5]
```diff
@@ -65,155 +51,101 @@ export const createNewUser = async (
...
-      await prisma.user.update({
-        where: {
-          id: user.id,
-        },
-        data: {
-          jwt: token,
-          isVerified: false,
-        },
-      });
+      if (
```
Comment (AarshShah9) [inline backend/controllers/user.controller.ts:110]: can we not use zod to validate this instead of doing a bunch of if/else checks

File: backend/controllers/user.controller.ts [hunk 6]
```diff
@@ -236,84 +168,38 @@ export const loginUser = async (
...
     }
   } catch (error) {
     next(error);
   }
 };
 
-// logout Student -> protected route
+// logout Student
 export const logoutUser = async (req: Request, res: Response) => {
-  const { email } = emailSchema.parse(req.body);
```
Comment (AarshShah9) [inline backend/controllers/user.controller.ts:305]: do we even need this route anymore? The front end just needs to purge the JWT token, no request is needed right? 

@jcharlpretorius @JamesRobert20

File: backend/prisma/schema.prisma [hunk 7]
```diff
@@ -66,16 +66,11 @@ model User {
 
   email       String @unique
   password    String @map("password")
-  yearOfStudy Int    @map("year_of_study")
```
Comment (AarshShah9) [inline backend/prisma/schema.prisma:69]: If you are getting rid of yearOfStudy please add it to the enrollment model

Comment (AarshShah9) [reply backend/prisma/schema.prisma:69]: And make sure to update the zodschemas as well for enrollment if there is any

File: shared/zodSchemas.ts [hunk 8]
```diff
@@ -369,11 +360,7 @@ export const InstitutionSchema = z.object({
     .min(3, { message: "Institution domain must at least 3 characters " }),
 });
 
-export const createInstitutionSchema = InstitutionSchema.omit({
-  id: true,
-  createdAt: true,
-  updatedAt: true,
-});
+export const createInstitutionSchema = InstitutionSchema.partial();
```
Comment (AarshShah9) [inline shared/zodSchemas.ts:363]: Why was this changed

File: backend/controllers/user.controller.ts [hunk 9]
```diff
@@ -65,155 +51,101 @@ export const createNewUser = async (
...
+        success: false,
+        message: "User already exists",
+      });
     }
 
-    const newUser = await prisma.user.create({
-      data: {
+    const token = jwt.sign(
+      {
+        institutionID: institution.id,
```
Comment (AarshShah9) [inline backend/controllers/user.controller.ts:62]: shouldn't it still be `institutionId`

File: backend/controllers/user.controller.ts [hunk 10]
```diff
@@ -65,155 +51,101 @@ export const createNewUser = async (
...
-    } else {
+        if (institution) {
+          const newUser = await prisma.user.create({
+            data: {
+              username: payload.username,
+              firstName: payload.firstName,
+              lastName: payload.lastName,
+              email: payload.email,
+              password: payload.password,
+              institutionId: payload.institutionID,
```
Comment (AarshShah9) [inline backend/controllers/user.controller.ts:134]: Reflect same requested changes above. Keep it consistent and use `institutionId`

File: backend/controllers/user.controller.ts [hunk 11]
```diff
@@ -65,155 +51,101 @@ export const createNewUser = async (
...
   try {
-    const { email, otp } = otpVerifySchema.parse(req.body);
-
-    const user = await prisma.user.findUnique({
-      where: {
-        email: email,
-        otp: otp,
-      },
-    });
+    const token = req.params.token;
```
Comment (AarshShah9) [inline backend/controllers/user.controller.ts:102]: please use zod validation

**Review by AarshShah9 — CHANGES_REQUESTED** [review]
Please tend to all requested changes

File: backend/controllers/user.controller.ts (hunk 1 above)
Comment (nomanSanjari) [reply backend/controllers/user.controller.ts:69]: Fixed

File: backend/controllers/user.controller.ts (hunk 2 above)
Comment (nomanSanjari) [reply backend/controllers/user.controller.ts:187]: fixed

File: backend/prisma/schema.prisma (hunk 3 above)
Comment (nomanSanjari) [reply backend/prisma/schema.prisma:73]: not sure why I did that, putting it back

File: backend/prisma/schema.prisma (hunk 4 above)
Comment (nomanSanjari) [reply backend/prisma/schema.prisma:49]: knew I was forgetting something

File: backend/controllers/user.controller.ts (hunk 6 above)
Comment (nomanSanjari) [reply backend/controllers/user.controller.ts:305]: yes it could do that. I'll leave it unchanged for now seeing as no one has said anything yet

File: shared/zodSchemas.ts (hunk 8 above)
Comment (nomanSanjari) [reply shared/zodSchemas.ts:363]: Zod was throwing an error even though the data was shaped properly and I didn't wanna spend any more time figuring it out so I threw this in to fix the issue to complete testing

File: backend/prisma/schema.prisma (hunk 7 above)
Comment (nomanSanjari) [reply backend/prisma/schema.prisma:69]: fixed

File: backend/controllers/user.controller.ts (hunk 9 above)
Comment (nomanSanjari) [reply backend/controllers/user.controller.ts:62]: fixed

File: backend/controllers/user.controller.ts (hunk 10 above)
Comment (nomanSanjari) [reply backend/controllers/user.controller.ts:134]: all instances of institutionID have been changed to institutionId

File: backend/controllers/user.controller.ts (hunk 5 above)
Comment (nomanSanjari) [reply backend/controllers/user.controller.ts:110]: fixed

File: backend/controllers/user.controller.ts (hunk 11 above)
Comment (nomanSanjari) [reply backend/controllers/user.controller.ts:102]: fixed

File: backend/controllers/user.controller.ts (hunk 6 above)
Comment (AarshShah9) [reply backend/controllers/user.controller.ts:305]: Okay sounds good, in the future we might even want to blacklist the JWT
